File: android/src/main/java/com/zenda/capacitor/httpserver/HttpServer.java
package com.zenda.capacitor.httpserver;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.content.Context;
import android.webkit.MimeTypeMap;
import com.getcapacitor.Logger;
import fi.iki.elonen.NanoHTTPD;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.ServerSocket;
import java.util.Map;

/**
 * HttpServer class manages the lifecycle and configuration of the local NanoHTTPD server on Android.
 */
public class HttpServer {

    // Instance of the internal NanoHTTPD server implementation
    private AndroidHttpServer server;
    // Cached URL where the server is accessible
    private String serverUrl;
    // The port number the server is listening on
    private int port;
    // The base directory from which static files are served
    private File baseDir;
    // The application context
    private Context context;
    // Flag to track user intent for the server state
    private boolean wasRunning = false;
    
    private static final int DEFAULT_STATIC_PORT = 55667;
    private static final String PREFS_NAME = "CapacitorHttpServerPrefs";
    private static final String PREF_PORT_KEY = "server_port";

    /**
     * Initializes the server configuration by reading the base directory from AndroidManifest meta-data.
     */
    public void init(Context context) {
        this.context = context;
        String basePath = "";
        try {
            // Retrieve application info to access meta-data from AndroidManifest.xml
            ApplicationInfo ai = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
            Bundle bundle = ai.metaData;
            // Check for the "HTTP_SERVER_BASE_DIR" configuration
            if (bundle != null && bundle.containsKey("HTTP_SERVER_BASE_DIR")) {
                basePath = bundle.getString("HTTP_SERVER_BASE_DIR");
            }
        } catch (PackageManager.NameNotFoundException e) {
            Logger.error("HttpServer", "Failed to load meta-data, NameNotFound: " + e.getMessage(), e);
        } catch (NullPointerException e) {
            Logger.error("HttpServer", "Failed to load meta-data, NullPointer: " + e.getMessage(), e);
        }

        // Set the base directory; default to app's internal files directory if not specified
        if (basePath != null && !basePath.isEmpty()) {
            this.baseDir = new File(context.getFilesDir(), basePath);
        } else {
            this.baseDir = context.getFilesDir();
        }
    }

    /**
     * Starts the HTTP server on an available port and returns its URL.
     */
    public String start() throws IOException {
        // If server is already running, return the existing URL
        if (server != null && server.isAlive()) {
            this.wasRunning = true;
            return serverUrl;
        }

        // 1. Try static port
        boolean started = tryStartOnPort(DEFAULT_STATIC_PORT);
        
        // 2. Try stored port if static port fails
        if (!started) {
            android.content.SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
            int storedPort = prefs.getInt(PREF_PORT_KEY, -1);
            if (storedPort != -1 && storedPort != DEFAULT_STATIC_PORT) {
                started = tryStartOnPort(storedPort);
            }
        }
        
        // 3. Try dynamic free port if all above failed
        if (!started) {
            int freePort = findFreePort();
            started = tryStartOnPort(freePort);
            if (started) {
                // Store the new dynamically chosen port
                android.content.SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
                prefs.edit().putInt(PREF_PORT_KEY, this.port).apply();
            }
        }

        if (!started) {
            throw new IOException("Failed to start server on any port.");
        }

        // Construct the server's local URL
        this.serverUrl = "http://localhost:" + port + "/";
        this.wasRunning = true;

        Logger.info("HttpServer", "Server started at " + serverUrl + " serving from " + baseDir.getAbsolutePath());
        return serverUrl;
    }
    
    private boolean tryStartOnPort(int targetPort) {
        try {
            this.server = new AndroidHttpServer(targetPort, baseDir);
            this.server.start(NanoHTTPD.SOCKET_READ_TIMEOUT, false);
            this.port = this.server.getListeningPort();
            return true;
        } catch (IOException e) {
            Logger.debug("HttpServer", "Failed to start server on port " + targetPort);
            this.server = null;
            return false;
        }
    }

    /**
     * Stops the running HTTP server and clears associated state.
     */
    public void stop() {
        this.wasRunning = false;
        if (server != null) {
            // Shut down the server
            server.stop();
            // Clear instance variables
            server = null;
            serverUrl = null;
            Logger.info("HttpServer", "Server stopped");
        }
    }
    
    /**
     * Used by app lifecycle hooks to pause the server without losing the user's intent.
     */
    public void pause() {
        if (server != null) {
            server.stop();
            server = null;
            Logger.info("HttpServer", "Server paused");
        }
    }
    
    public boolean wasRunning() {
        return this.wasRunning;
    }

    /**
     * Returns the current server URL.
     */
    public String getUrl() {
        return serverUrl;
    }

    /**
     * Returns the active status and connection info as a JSObject.
     */
    public com.getcapacitor.JSObject getActiveStatus() {
        com.getcapacitor.JSObject status = new com.getcapacitor.JSObject();
        boolean active = (server != null && server.isAlive());
        status.put("active", active);
        if (active) {
            status.put("port", this.port);
            status.put("hostname", "localhost");
            status.put("protocol", "http:");
            status.put("url", this.serverUrl);
        }
        return status;
    }

    /**
     * Finds an available port on the device by opening a ServerSocket on port 0.
     */
    private int findFreePort() {
        try (ServerSocket socket = new ServerSocket(0)) {
            return socket.getLocalPort();
        } catch (IOException e) {
            // Fallback port if automatic detection fails
            return 8080;
        }
    }

    /**
     * Internal implementation of NanoHTTPD to serve files from the specified base directory.
     */
    private static class AndroidHttpServer extends NanoHTTPD {

        private final File baseDir;

        public AndroidHttpServer(int port, File baseDir) {
            super(port);
            this.baseDir = baseDir;
        }

        @Override
        public Response serve(IHTTPSession session) {
            // Extract the requested URI
            String uri = session.getUri();
            
            // Map the URI to a file within the base directory
            // Note: uri starts with /
            File file = new File(baseDir, uri.substring(1));

            try {
                // Security: Prevent Path Traversal attacks
                // Ensure the resolved canonical path is still within the base directory
                String canonicalBase = baseDir.getCanonicalPath();
                String canonicalFile = file.getCanonicalPath();
                
                if (!canonicalFile.startsWith(canonicalBase)) {
                    // Logger.error("HttpServer", "Blocked path traversal attempt: " + uri);
                    return newFixedLengthResponse(Response.Status.FORBIDDEN, NanoHTTPD.MIME_PLAINTEXT, "Forbidden: Path traversal attempt");
                }

                // If the file exists and is indeed a file, serve its content
                if (file.exists() && file.isFile()) {
                    // Determine MIME type using Android's native MimeTypeMap
                    String mimeType = getMimeType(uri);
                    InputStream inputStream = new FileInputStream(file);
                    // Return a chunked response for the file content
                    return newChunkedResponse(Response.Status.OK, mimeType, inputStream);
                }
            } catch (IOException e) {
                // Return internal error if file reading or path resolution fails
                Logger.error("HttpServer", "Error serving file: " + uri, e);
                return newFixedLengthResponse(
                    Response.Status.INTERNAL_ERROR,
                    NanoHTTPD.MIME_PLAINTEXT,
                    "Error processing request: " + e.getMessage()
                );
            }

            // Return 404 Not Found if the file doesn't exist or is outside baseDir
            return newFixedLengthResponse(Response.Status.NOT_FOUND, NanoHTTPD.MIME_PLAINTEXT, "File not found: " + uri);
        }

        /**
         * Resolves MIME type using Android's native MimeTypeMap.
         */
        private String getMimeType(String uri) {
            String extension = MimeTypeMap.getFileExtensionFromUrl(uri);
            if (extension != null) {
                String type = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension.toLowerCase());
                if (type != null) {
                    return type;
                }
            }
            // Fallback to plaintext
            return NanoHTTPD.MIME_PLAINTEXT;
        }
    }
}

File: ios/Sources/HttpServerPlugin/HttpServer.swift
import Foundation
import GCDWebServer
import UniformTypeIdentifiers
import Capacitor
import MobileCoreServices

/// The HttpServer class manages the lifecycle and configuration of the local GCDWebServer.
@objc public class HttpServer: NSObject {
    // Instance of the GCDWebServer
    private var webServer: GCDWebServer?
    // Cached URL of the running server
    private var serverUrl: String?
    // Base directory from which files will be served
    private var baseDir: URL?
    
    // User intent state for lifecycle management
    public var wasRunning = false
    
    private let defaultStaticPort = 55667
    private let portKey = "CapacitorHttpServerPort"

    /// Initializes the server configuration, determining the base directory for serving files.
    @objc public func initialize() {
        let fileManager = FileManager.default
        // Get the app's documents directory as the default root
        let documentsURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!
        
        var basePath = ""
        // Check Info.plist for a custom base directory configuration
        if let path = Bundle.main.object(forInfoDictionaryKey: "HTTP_SERVER_BASE_DIR") as? String {
            basePath = path
        }
        
        // If a base path is configured, append it to the documents URL
        if !basePath.isEmpty {
            self.baseDir = documentsURL.appendingPathComponent(basePath)
        } else {
            // Otherwise, use the documents directory directly
            self.baseDir = documentsURL
        }
    }

    /// Starts the HTTP server and returns the server URL.
    @objc public func start() -> String? {
        // Return existing URL if server is already running
        if let url = serverUrl, webServer?.isRunning == true {
            self.wasRunning = true
            return url
        }

        // Ensure the base directory is set before starting
        guard let baseDir = self.baseDir else {
            return nil
        }

        // Stop any existing inactive server before we reassign
        webServer?.stop()
        webServer = nil
        
        // 1. Try static port
        var startedServer = startServerInstance(on: defaultStaticPort, baseDir: baseDir)
        
        // 2. Try stored port if static port fails
        if startedServer == nil {
            let storedPort = UserDefaults.standard.integer(forKey: portKey)
            if storedPort > 0 && storedPort != defaultStaticPort {
                startedServer = startServerInstance(on: storedPort, baseDir: baseDir)
            }
        }
        
        // 3. Try dynamic free port if all above failed
        if startedServer == nil {
            startedServer = startServerInstance(on: 0, baseDir: baseDir) // dynamic port fallback
            if let server = startedServer {
                UserDefaults.standard.set(Int(server.port), forKey: portKey)
            }
        }

        if let server = startedServer, let url = server.serverURL {
            self.webServer = server
            self.serverUrl = url.absoluteString
            self.wasRunning = true
            return self.serverUrl
        }

        return nil
    }
    
    private func startServerInstance(on port: Int, baseDir: URL) -> GCDWebServer? {
        let server = GCDWebServer()
        
        // Use a weak reference to self in the handler block to avoid retain cycles
        weak var weakSelf = self
        
        // Add a handler to serve files with security checks and native MIME type resolution
        server.addDefaultHandler(forMethod: "GET", request: GCDWebServerRequest.self, processBlock: { request in
            guard let self = weakSelf else {
                return GCDWebServerErrorResponse(statusCode: 500)
            }
            
            let path = request.path
            // Map the request path to the absolute file path
            let fileURL = baseDir.appendingPathComponent(path.hasPrefix("/") ? String(path.dropFirst()) : path)
            
            // Security: Prevent Path Traversal attacks
            // Ensure the resolved file path is still within the base directory
            let canonicalFile = fileURL.resolvingSymlinksInPath().path
            let canonicalBase = baseDir.resolvingSymlinksInPath().path
            
            if !canonicalFile.hasPrefix(canonicalBase) {
                CAPLog.print("Blocked path traversal attempt: \(path)")
                return GCDWebServerErrorResponse(statusCode: 403)
            }
            
            // Serve the file if it exists and is not a directory, or serve index.html if it is a directory
            var isDirectory: ObjCBool = false
            if FileManager.default.fileExists(atPath: fileURL.path, isDirectory: &isDirectory) {
                if isDirectory.boolValue {
                    let indexFileURL = fileURL.appendingPathComponent("index.html")
                    var indexIsDirectory: ObjCBool = false
                    if FileManager.default.fileExists(atPath: indexFileURL.path, isDirectory: &indexIsDirectory), !indexIsDirectory.boolValue {
                        let mimeType = self.getMimeType(for: indexFileURL)
                        let response = GCDWebServerFileResponse(file: indexFileURL.path)
                        response?.contentType = mimeType
                        return response
                    }
                } else {
                    // Determine MIME type using iOS native UniformTypeIdentifiers
                    let mimeType = self.getMimeType(for: fileURL)
                    let response = GCDWebServerFileResponse(file: fileURL.path)
                    response?.contentType = mimeType
                    return response
                }
            }
            
            // SPA Fallback: walk up the directory tree to find an index.html
            var fallbackDir = fileURL.deletingLastPathComponent()
            
            while fallbackDir.path.hasPrefix(baseDir.path) {
                let indexFileURL = fallbackDir.appendingPathComponent("index.html")
                var indexIsDirectory: ObjCBool = false
                if FileManager.default.fileExists(atPath: indexFileURL.path, isDirectory: &indexIsDirectory), !indexIsDirectory.boolValue {
                    let mimeType = self.getMimeType(for: indexFileURL)
                    let response = GCDWebServerFileResponse(file: indexFileURL.path)
                    response?.contentType = mimeType
                    return response
                }
                
                if fallbackDir.path == baseDir.path {
                    break
                }
                fallbackDir = fallbackDir.deletingLastPathComponent()
            }
            
            // Return 404 if file not found
            return GCDWebServerErrorResponse(statusCode: 404)
        })
        
        let options: [String: Any] = [
            GCDWebServerOption_Port: port,
            GCDWebServerOption_BindToLocalhost: true,
            GCDWebServerOption_AutomaticallySuspendInBackground: false
        ]
        
        do {
            try server.start(options: options)
            return server
        } catch {
            // CRITICAL: Stop the server to clean up internal _options if start failed
            // This prevents a crash when the GCDWebServer instance is deallocated
            server.stop()
            return nil
        }
    }

    /// Stops the running HTTP server and clears associated state.
    @objc public func stop() {
        self.wasRunning = false
        // Stop the server if it's running
        webServer?.stop()
        // Reset server instance and URL cache
        webServer = nil
        serverUrl = nil
    }

    /// Pauses the running HTTP server for backgrounding without modifying user intent.
    @objc public func pause() {
        webServer?.stop()
        webServer = nil
        serverUrl = nil
    }

    /// Returns the current server URL if the server is running.
    @objc public func getUrl() -> String? {
        return serverUrl
    }
    
    /// Returns the active status and connection properties.
    @objc public func getActiveStatus() -> [String: Any] {
        let active = webServer?.isRunning == true
        var status: [String: Any] = ["active": active]
        if active {
            status["port"] = webServer?.port
            status["hostname"] = "localhost"
            status["protocol"] = "http:"
            status["url"] = self.serverUrl
        }
        return status
    }
    
    /// Resolves MIME type for a given URL using native system APIs.
    private func getMimeType(for url: URL) -> String {
        if #available(iOS 14.0, *) {
            if let type = UTType(filenameExtension: url.pathExtension),
               let mimeType = type.preferredMIMEType {
                return mimeType
            }
        } else {
            // Fallback for older iOS versions
            if let ident = UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, url.pathExtension as CFString, nil as CFString?)?.takeRetainedValue(),
               let type = UTTypeCopyPreferredTagWithClass(ident, kUTTagClassMIMEType)?.takeRetainedValue() as String? {
                return type
            }
        }
        return "application/octet-stream"
    }
}
